keps/sig-node/20190129-hugepages.md

```diff
@@ -279,8 +290,94 @@ opencontainers/runtime-spec.
 see:
 https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#huge-page-limits
 
+The runtime-spec says `hugepageLimits` as an array of objects, and each object
+allows specifying `limit` per `pageSize`.
+```
+    "hugepageLimits": [
+        {
+            "pageSize": "2MB",
+            "limit": 209715200
+        },
+        {
+            "pageSize": "64KB",
+            "limit": 1000000
+        }
+   ]
+```
+
+The `LinuxContainerResources` message can be extended to specify multiple size
+to align with the runtime-spec.
+```
+message LinuxContainerResources {
+    ...
+    string cpuset_mems = 7;
+    // List of Hugepage limit, the limit consists of pagesize and limit.
+    repeated HugepageLimit hugepage_limits = 8;
+}
+
+// HugepageLimit contains pageSize and limit.
+// `page_size=1GB`, `limit=1073741824` means setting `1073741824` bytes
+// on hugetlb.1GB.limit_in_bytes of container level cgroup.
+message HugepageLimit {
+    // hugepage size in the format <size><unit-prefix>B (2MB, 1GB)
+    string page_size = 1;
+    // hugepages limit in bytes.
+    uint64 limit = 2;
+}
+```
+
 The CRI changes are required before promoting this feature to Beta.
 
+#### Support container isolation of huge pages
+
+Container isolation of hugepages should be supported to guarantee containers to
+consume huge pages as requested and to support NUMA in the future.  To support
+container isolation, the `LinuxContainerResources` message, which is used by the
+`UpdateContainerResources` RPC of CRI, should be extended to support huge pages.
+
+At this point(Kubernetes 1.16), Kubelet calculates the total requested amount of
+huge pages from the containers in the pod spec then sets the aggregated limits
+of huge pages on the pod level cgroup directly.
+
+The `Hugepage Handler` will be introduced to support container isolation of
```

Comment:
I do not understand why we need a dedicated handler.

I would have anticipated that when doing the implementation we would just integrate this with kuberuntime_container_linux so when we generateLinuxContainerConfig we would just set the hugepage limitations at the same time we set the cpu shares, etc.

I do not see why we need to update container resources rather than just set them initially when creating the container.  In general, this is a detail we can work through in the implementation phase.

Reply:
Agreed,  it is reasonable to set hugepages over `LinuxContainerConfig` CRI message, as you suggested.

So, the approach will be changed to set hugepages limit on `LinuxContainerConfig` message when `generateLinuxContainerConfig` called.

```
// LinuxContainerConfig contains platform-specific configuration for
// Linux-based containers.
message LinuxContainerConfig {
    // Resources specification for the container.
    LinuxContainerResources resources = 1;
    // LinuxContainerSecurityContext configuration for the container.
    LinuxContainerSecurityContext security_context = 2;
}
```